modules/auxiliary/gather/asrep.rb

```diff
@@ -0,0 +1,157 @@
+##
+# This module requires Metasploit: https://metasploit.com/download
+# Current source: https://github.com/rapid7/metasploit-framework
+##
+
+class MetasploitModule < Msf::Auxiliary
+  include Msf::Exploit::Remote::Kerberos::Client
+  include Msf::Exploit::Remote::LDAP
+  include Msf::Exploit::Remote::LDAP::Queries
+
+  def initialize(info = {})
+    super(
+      update_info(
+        info,
+        'Name' => 'Find Users Without Pre-Auth Required (ASREP-roast)',
+        'Description' => %q{
+          This module searches for AD users without pre-auth required. Two different approaches
+          are provided:
+          - Brute force of usernames (does not require a user account; should not lock out accounts)
+          - LDAP lookup (requires an AD user account)
+        },
+        'License' => MSF_LICENSE,
+        'Author' => [
+          'smashery', # MSF Module
+        ],
+        'References' => [
+          ['URL', 'https://www.ired.team/offensive-security-experiments/active-directory-kerberos-abuse/as-rep-roasting-using-rubeus-and-hashcat']
+        ],
+        'Notes' => {
+          'Stability' => [CRASH_SAFE],
+          'SideEffects' => [IOC_IN_LOGS],
+          'Reliability' => [],
+          'AKA' => ['preauth', 'asreproast']
+        },
+        'Actions' => [
+          ['BRUTE_FORCE', { 'Description' => 'Brute force to find susceptible user accounts' } ],
+          ['LDAP', { 'Description' => 'Ask a domain controller directly for the susceptible user accounts' } ],
+        ],
+        'DefaultAction' => 'BRUTE_FORCE'
+      )
+    )
+
+    register_options(
+      [
+        OptPath.new('USER_FILE', [ false, 'File containing usernames, one per line' ], conditions: %w[ACTION == BRUTE_FORCE]),
+        OptBool.new('USE_RC4_HMAC', [ true, 'Request using RC4 hash instead of default encryption types (faster to crack)', true]),
+        OptString.new('Rhostname', [ true, "The domain controller's hostname"], aliases: ['LDAP::Rhostname']),
+      ]
+    )
+    register_advanced_options(
+      [
+        OptEnum.new('LDAP::Auth', [true, 'The Authentication mechanism to use', Msf::Exploit::Remote::AuthOption::NTLM, Msf::Exploit::Remote::AuthOption::LDAP_OPTIONS]),
```

Comment:
Does this have to be registered because there's a problem when it's left at its default value of `auto`?

Reply:
Yeah, if it's auto, it just uses plaintext, which isn't supported on domain controllers, so it just fails. Given this won't be run on non-DC LDAP servers, I think that's a reasonable default.